Implement QubitUnitary's decomposition under the new system (#7211)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

> from pennylane import QubitUnitary:orphan:

Reviewer: ```suggestion
:orphan:
```

  gate-set that can be translated to the MBQC formalism.
  [(7271)](https://github.com/PennyLaneAI/pennylane/pull/7271)

* Added `convert_to_su2` and `convert_to_su4` to `qml.math` that converts a unitary matrix to SU(2)
  and SU(4) respectively, and optionally a global phase.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

<h4>Resource-efficient Decompositions 🔎</h4>

* The new graph-based decomposition system now supports `QubitUnitary`.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

* Added `qml.decomposition.DecompositionNotApplicable` that can be raised from decomposition rules
  if the decomposition rule is not applicable under a certain set of parameters.

>   [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

Author: @isaacdevlugt let me know if we decide on a different API for conditionally applicable decomposition rules.

Reviewer 2: Thanks @astralcai! I think this is fine for now... it's a little clunky looking, but if we have good examples to supplement it with, this will help!

  can be accessed with the new graph-based decomposition system. The most efficient set of rotations to 
  decompose into will be chosen based on the target gate set.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

  ```python
  from functools import partial
  import numpy as np
  import pennylane as qml
  
  qml.decomposition.enable_graph()
  
  U = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
  
  @partial(qml.transforms.decompose, gate_set={"RX", "RY", "GlobalPhase"})
  @qml.qnode(qml.device("default.qubit"))
  def circuit():
      qml.QubitUnitary(np.array([[1, 1], [1, -1]]) / np.sqrt(2), wires=[0])
      return qml.expval(qml.PauliZ(0))
  ```
  ```pycon
  >>> print(qml.draw(circuit)())
  0: ──RX(0.00)──RY(1.57)──RX(3.14)──GlobalPhase(-1.57)─┤  <Z>
  ```

* Decomposition rules can be marked as not-applicable with `qml.decomposition.DecompositionNotApplicable`, allowing for flexibility when creating conditional decomposition 
  rules based on parameters that affects the rule's resources.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

  ```python

>   from pennylane import qml

Reviewer: ```suggestion
  import pennylane as qml
```
LMAO

  gate-set that can be translated to the MBQC formalism.
  [(7271)](https://github.com/PennyLaneAI/pennylane/pull/7271)

> * Two new functions called `convert_to_su2` and `convert_to_su4` have been added to `qml.math`, which convert unitary matrices to SU(2) or SU(4), respectively, and optionally a global phase.

Reviewer 2: Can we add RTD links to these?

Author: Done

> * Added `convert_to_su2` and `convert_to_su4` to `qml.math` that converts a unitary matrix to SU(2)
>   and SU(4) respectively, and optionally a global phase.
>   [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

Reviewer 2: ```suggestion
* Two new functions called `convert_to_su2` and `convert_to_su4` have been added to `qml.math`, which convert unitary matrices to SU(2) or SU(4), respectively, and optionally a global phase.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)
```

<h4>Resource-efficient Decompositions 🔎</h4>

* New decomposition rules comprising rotation gates and global phases have been added to `QubitUnitary` that 
  can be accessed with the new graph-based decomposition system. The most efficient set of rotations to 
  decompose into will be chosen based on the target gate set.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

  ```python
  from functools import partial
  import numpy as np
  import pennylane as qml
  
  qml.decomposition.enable_graph()
  
  U = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
  
  @partial(qml.transforms.decompose, gate_set={"RX", "RY", "GlobalPhase"})
  @qml.qnode(qml.device("default.qubit"))
  def circuit():
      qml.QubitUnitary(np.array([[1, 1], [1, -1]]) / np.sqrt(2), wires=[0])
      return qml.expval(qml.PauliZ(0))
  ```
  ```pycon
  >>> print(qml.draw(circuit)())
  0: ──RX(0.00)──RY(1.57)──RX(3.14)──GlobalPhase(-1.57)─┤  <Z>
  ```

> * Decomposition rules can be marked as not-applicable with `qml.decomposition.DecompositionNotApplicable`, allowing for flexibility when creating conditional decomposition 

Reviewer 2: Can we add an RTD link to this as well?

Author: Done

* Added `qml.decomposition.DecompositionNotApplicable` that can be raised from decomposition rules
  if the decomposition rule is not applicable under a certain set of parameters.
  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

  ```python
  from pennylane import qml
  from pennylane.decomposition import DecompositionNotApplicable
  from pennylane.math.decomposition import zyz_rotation_angles
  
  def _zyz_resource(num_wires):
      if num_wires != 1:
          # This decomposition is only applicable when num_wires is 1
          raise DecompositionNotApplicable
      return {qml.RZ: 2, qml.RY: 1, qml.GlobalPhase: 1}

  def zyz_decomposition(U, wires, **__):
      phi, theta, omega, phase = zyz_rotation_angles(U, return_global_phase=True)
      qml.RZ(phi, wires=wires[0])
      qml.RY(theta, wires=wires[0])
      qml.RZ(omega, wires=wires[0])
      qml.GlobalPhase(-phase)

>   ```

Author: This actually doesn't work, because `wires` is not in the `resource_keys` of `CNOT`.

>   This decomposition will be ignored for `QubitUnitary` on more than one wire.

Reviewer 2: Probably just missing an `add_decomps(QubitUnitary, ...)` above this

Author: Done

<h4>Resource-efficient Decompositions 🔎</h4>

> * The new graph-based decomposition system now supports `QubitUnitary`.

Reviewer 2: Can we be a bit more specific regarding the user impact here? E.g., 

> New decomposition rules have been added to `QubitUnitary` that can be accessed with the graph-based decomposition system. <sentence on the benefit of these decomposition rules (what can the user do now that they couldn't before?)>

Wouldn't hurt to have a code example here, too :). Even if it's just printing entries in `list_decomps`, say!

  [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

* Added `qml.decomposition.DecompositionNotApplicable` that can be raised from decomposition rules
  if the decomposition rule is not applicable under a certain set of parameters.

>   [(#7211)](https://github.com/PennyLaneAI/pennylane/pull/7211)

Reviewer 2: A small example would be nice here 👀

Author: Done
